### ISSUE TITLE
RobolectricTestRunner frozen randomly

### ISSUE THREAD

**Issue by gmerinojimenez** [robolectric/robolectric#2049]
Hi,

We're having random issues when executing Robolectric tests. 

Main thread gets stucked on getShadowClassName and it's unable to continue the execution.

The test where this happens is never the same, neither the portion of app code where it's triggered. We're trying to checkout several revisions to check if we're introducing this problem at some point in our code but we weren't able to find it.

Here is the main thread dump for one of the frozen tests:

```
"Test worker" prio=5 tid=0x00007fe88b1a5000 nid=0x5503 runnable [0x000000010e911000]
   java.lang.Thread.State: RUNNABLE
	at java.util.HashMap.getEntry(HashMap.java:465)
	at java.util.HashMap.get(HashMap.java:417)
	at org.robolectric.internal.bytecode.ShadowWrangler.getShadowConfig(ShadowWrangler.java:171)
	at org.robolectric.internal.bytecode.ShadowWrangler.getShadowClassName(ShadowWrangler.java:385)
	at org.robolectric.internal.bytecode.ShadowWrangler.createShadowFor(ShadowWrangler.java:366)
	at org.robolectric.internal.bytecode.ShadowWrangler.initializing(ShadowWrangler.java:106)
	at org.robolectric.internal.bytecode.RobolectricInternals.initializing(RobolectricInternals.java:18)
	at android.util.TypedValue.$$robo$init(TypedValue.java)
	at android.util.TypedValue.<init>(TypedValue.java)
	at com.android.internal.policy.impl.PhoneWindow.$$robo$$__constructor__(PhoneWindow.java:141)
	at com.android.internal.policy.impl.PhoneWindow.<init>(PhoneWindow.java)
	at org.robolectric.shadows.ShadowWindow.create(ShadowWindow.java:27)
	at sun.reflect.GeneratedMethodAccessor220.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.robolectric.util.ReflectionHelpers.callStaticMethod(ReflectionHelpers.java:226)
	at org.robolectric.internal.bytecode.ShadowWrangler$5.call(ShadowWrangler.java:263)
	at org.robolectric.internal.bytecode.ShadowWrangler.intercept(ShadowWrangler.java:227)
	at org.robolectric.internal.bytecode.RobolectricInternals.intercept(RobolectricInternals.java:33)
	at android.app.Activity.$$robo$$attach(Activity.java:5876)
	at android.app.Activity.attach(Activity.java)
	at sun.reflect.GeneratedMethodAccessor225.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
	at org.robolectric.internal.runtime.Api21RuntimeAdapter.callActivityAttach(Api21RuntimeAdapter.java:20)
	at org.robolectric.util.ActivityController.attach(ActivityController.java:56)
	at org.robolectric.util.ActivityController$1.run(ActivityController.java:121)
	at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:304)
	at org.robolectric.shadows.CoreShadowsAdapter$2.runPaused(CoreShadowsAdapter.java:45)
	at org.robolectric.util.ActivityController.create(ActivityController.java:118)
	at org.robolectric.util.ActivityController.create(ActivityController.java:129)
	at com.mycompany.messenger.video.ui.screen.PlayUserVideoFragmentIT.attachToActivityAndStart(PlayUserVideoFragmentIT.java:175)
	at com.mycompany.messenger.video.ui.screen.PlayUserVideoFragmentIT.shouldShowProgressBarWhenLoadingRemoteVideo(PlayUserVideoFragmentIT.java:91)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:251)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:188)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:152)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at sun.reflect.GeneratedMethodAccessor256.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.GeneratedMethodAccessor255.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

Thank you for your time

**Comment by ndahlquist** [comment]
We have also seen this. It's fairly infrequent (maybe 1 out of 30 builds). Our thread dumps show the same stack trace as above, which seems odd because the thread is still runnable and there doesn't seem to be a reason why the execution should stop. Any ideas?

**Comment by dpastor** [comment]
Hi. We started also with a few builds failing because of this, but we ended on having like 1 out of 3  builds failing because of it. With the many stacktraces we were able to obtain, we determined that the common stacktrace is:
```
   java.lang.Thread.State: RUNNABLE
    at java.util.HashMap.getEntry(HashMap.java:465)
    at java.util.HashMap.get(HashMap.java:417)
    at org.robolectric.internal.bytecode.ShadowWrangler.getShadowConfig(ShadowWrangler.java:171)
```

Taking a look on similar issues:
https://jira.nuxeo.com/browse/NXP-15254
http://stackoverflow.com/questions/31994559/hibernate-stuck-in-hashmap-getentry

And that, in fact, getShadowConfig can be called from multiple threads, it's seems just a concurrency problem with the HashMap.

After replacing shadowConfigCache HashMap by a ConcurrentHashMap, we are not able to reproduce the problem anymore.

I'm preparing a pull-request for this.

**Comment by chariotsystems** [comment]
Whilst this is fixed on 3.1-SNAPSHOT there are a number of reasons why you would not want to move off 3.0.  These are:
A number of the APIs have changed specifically:
 SharedPreferences preferences = ShadowPreferenceManager.getDefaultSharedPreferences(app);

Secondly 3.1 will not run against sdk 23 because it drags in version 18 of shadow-core which is no longer available in any of the maven repos.

I have found a workaround which allows you to use 3.0 and still have the fix that works:

public class CorporateTestRunner extends RobolectricGradleTestRunner {

    public CorporateTestRunner(Class<?> testClass) throws InitializationError {
        super(testClass);
    }

    protected ClassHandler createClassHandler(ShadowMap shadowMap, SdkConfig sdkConfig) {
        ClassHandler classHandler = new ShadowWrangler(shadowMap);
        try {
            Field field = classHandler.getClass().getDeclaredField("shadowConfigCache");
            field.setAccessible(true);
            field.set(classHandler, new ConcurrentHashMap<Class, ShadowConfig>());
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
        return classHandler;
    }

...
}
